### DISCUSSION TITLE
Apprise-API attachments | Confused 😕

### DISCUSSION THREAD

**Discussion by flinthamm** [caronc/apprise-api#140] [General]
OK, I'm sorry but I'm confused and hoping someone can help out?

I'm predominantly testing Apprise with Telegram, ntfy.sh and email. If I send direct to these services they support rich push messaging of various degrees but can included text, html, attachments, in-line images, etc. 

I wrongly assumed and tried for a long time to use the Apprise-API through postman with a working stateful (YAML) config but with attachments and could not get them working. The messages would send but no attachment would appear. After enabling DEBUG and more verbose logging I still could not find the key, together with adding more config (this is the latest LinuxServer IO Docker Image v0.9.5-ls97 Oct'23)  which should underlying be using AppRise v1.6.0. 

I then found documentation saying the Apprise-Api did not support attachments. Then I found Github Pull [#118](https://github.com/caronc/apprise-api/pull/118) that suggested this had been added, together with the Reddit post [Apprise v1.4.0 Released - Apprise API Attachment Support](https://www.reddit.com/r/selfhosted/comments/13im1mr/apprise_v140_released_apprise_api_attachment/#:~:text=One%20of%20the%20big%20features,with%20the%20Apprise%20Python%20API.).

However a simple curl call either with normal form-data fields or a JSON Body, see below delivers but without any errors and no attachment. 
```
POST http://localip:8000/notify/myconfig 
{"tag":"mytag-goes-here", "title":"My title goes here", "body":"Test message", "attach":"https://mydomain.com/myimage.jpg"}

```
I suspect someone is going to say this has not been added (I did seem some mention of security concerns, at least for external attachments, although you can do this via direct calls to the relevant services API so did not really understand.). Perhaps I'm misunderstanding Apprise API with the Web API or something else?

Anyway, hope this will clear everything up for anyone else with the same question as the documentation does not seem to make this clear anywhere. The services in question are also listed in the documentation and show as "Attahcments" supported.

**Comment by caronc** [comment]
So if everything working for you except attachments.

> I then found documentation saying the Apprise-Api did not support attachments.

It definitely supports attachments


> However a simple curl call either with normal form-data fields or a JSON Body, see below delivers but without any errors and no attachment. 
> ```
> POST http://localip:8000/notify/myconfig 
> {"tag":"mytag-goes-here", "title":"My title goes here", "body":"Test message", "attach":"https://mydomain.com/myimage.jpg"}
> 
> ```

You need you use the `-T` to add the attachment if you're using `curl`.

If your use the ApI (website) to test, do attachments work then? You can use the browse button and add attachments that way to test it out

**Reply by flinthamm** [reply]
@caronc Thank you for your reply and confirmation that's great news. Unfortauntely, I still can't get this to work using the API website to test, see below:
![image](https://github.com/caronc/apprise-api/assets/1732958/14515d8b-9122-4ca5-82df-050e13266ef5)

Using Postman doesn't improve the situation either, I will just test with Curl and the -T (upload flag) which I suspect might. If we want to send a push message from a security camera (for instance) I was hoping to just specify the still image capture url as the attachments (either http:// or https://) to get this to appear in a push message.
![image](https://github.com/caronc/apprise-api/assets/1732958/3e0bd8a6-cf12-4322-a82f-a88882abbaf6)
The above works but no attachment appears in the push message when sending (via tags) to Telegram, ntfy or Email?

![image](https://github.com/caronc/apprise-api/assets/1732958/60b973b5-0d9f-4784-8b16-f4934bf53300)
Choosing the actual binary file in postman rather than specifying a url path, much like the Apprise web interface, just produces a "Bad attachment" error?

Hoping you could shed some light on this - Thanks once again, this is much appreciated.

**Comment by flinthamm** [comment]
If I try curl, I assume from the error you can't send other multi-part fields at the same time as using the `-T` flag, is that correct? If so, how can you direct the attachment to a specific service if you can not specify the tag?
![image](https://github.com/caronc/apprise-api/assets/1732958/bd617ab9-b782-4a27-86d4-63181532bd29)

I've tried removing all the other fields `-F` but I'm now getting another error `URL bad/illegal format or missing URL message`, although this may just be a curl issue as I'm using Windows. Anyway this is the progress so far.

**Reply by caronc** [reply]
That is unfortunate. It's possible i broke the attachment feature in the last release, but i won't be able to confirm until i get home and have laptop access next week.

I appologize. My thought process is: If it doesn't work from the gui (using browse button), i very much suspect something is wrong on my end.

Worst case, you could use the tag from the previous docker version for the time being to see if that works?

**Comment by flinthamm** [comment]
Just to also confirm for testing I've given the `./attach` volume mount full rw permissions and confirmed by connecting a shell to the running container that files can be written to and read from this directory.

Sorry, just got your response as I was typing. I'll try this ASAP and let you know.

**Reply by flinthamm** [reply]
I've run out of time for now but have had chance to pull down the last release tag `0.9.4` and via the web interface, this immediately displayed a VERY long error message (as DEBUG is enabled). A short version of the message is below but suggests a permission error, despite having just `sudo chmod 777` the ./attach directory !?

```
ValueError at /notify/apprise
Could not create directory /app/apprise-api/apprise_api/var/attach
Request Method:	POST
Request URL:	http://<redacted-ip>/notify/apprise
Django Version:	4.2.6
Exception Type:	ValueError
Exception Value:	
Could not create directory /app/apprise-api/apprise_api/var/attach
Exception Location:	/app/apprise-api/apprise_api/api/utils.py, line 84, in __init__
Raised during:	api.views.NotifyView
Python Executable:	/usr/sbin/uwsgi
Python Version:	3.11.6
Python Path:	
['.',
 '',
 '/usr/lib/python311.zip',
 '/usr/lib/python3.11',
 '/usr/lib/python3.11/lib-dynload',
 '/lsiopy/lib/python3.11/site-packages']
Server time:	Fri, 20 Oct 2023 04:38:25 -0500
Traceback Switch to copy-and-paste view
/app/apprise-api/apprise_api/api/utils.py, line 80, in __init__
    def __init__(self, filename, path=None, delete=True):
        """
        Initialize our attachment
        """
        self._filename = filename
        try:
            os.makedirs(settings.APPRISE_ATTACH_DIR, exist_ok=True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ …
        except OSError:
            # Permission error
            raise ValueError('Could not create directory {}'.format(
                settings.APPRISE_ATTACH_DIR))
Local vars
<frozen os>, line 225, in makedirs
<source code not available>
             ^^^^^^^^^^^^^^^^^ …
Local vars
During handling of the above exception ([Errno 13] Permission denied: '/app/apprise-api/apprise_api/var/attach'), another exception occurred:
/lsiopy/lib/python3.11/site-packages/django/core/handlers/exception.py, line 55, in inner
        return inner
    else:
        @wraps(get_response)
        def inner(request):
            try:
                response = get_response(request)
                               ^^^^^^^^^^^^^^^^^^^^^ …
            except Exception as exc:
                response = response_for_exception(request, exc)
            return response
        return inner
Local vars
```
I'll do some more testing as time allows, although the permission error seems strange. Thanks for your help thus far.

**Reply by caronc** [reply]
Maybe just `chmod 777` (when you have time) your externally mounted `/attach` dir for now? Or just don't expose it (from the docker instance) as a test since its permissions will be set fine internally.

Good luck

**Comment by caronc** [comment]
Bad Attachment issue has now been resolved in #236
